### PR TITLE
Progressive release cancelation UI

### PR DESCRIPTION
**Pull request by Lukewh**
Fixes https://github.com/canonical-web-and-design/snapcraft.io/issues/2342

## QA
- Pull branch
- `./run`
- Visit http://0.0.0.0:8004/snap_name/releases
- If you don't have a progressive release - create one and save
- Cancel the progressive release
- See a new item in the save bar regarding "cancelling" a progressive release

### CODE REVIEW

File: static/js/publisher/release/actions/pendingReleases.test.js [hunk 1]
```diff
@@ -225,4 +227,27 @@ describe("pendingReleases actions", () => {
       expect(resumeProgressiveRelease(key).payload).toBe(key);
     });
   });
+
+  describe("cancelProgressiverelease", () => {
+    const key = "progressive-test";
+    const previousRevision = "previous-revision";
```
Comment (bartaz) [inline static/js/publisher/release/actions/pendingReleases.test.js]: Would be good for this to be the same as in real case.
Is `previousRevision` supposed to be revision id? Whole revision object? This test doesn't tell which one.

File: static/js/publisher/release/reducers/pendingReleases.js [hunk 2]
```diff
@@ -154,6 +155,27 @@ function resumeProgressiveRelease(state, key) {
   return nextState;
 }
 
+function cancelProgressiveRelease(state, key, previousRevision) {
```
Comment (bartaz) [inline static/js/publisher/release/reducers/pendingReleases.js]: Is this supposed to cancel all pending revisons with the same key?

What would happen if several revisions across different architectures (for example whole channel) was released progressively with the same key.
If `cancelProgressiveRelease` cancels all releases with the same key, will it cancel them all? How can only be just one `previousRevision`?

File: static/js/publisher/release/reducers/pendingReleases.js [hunk 3]
```diff
@@ -154,6 +155,27 @@ function resumeProgressiveRelease(state, key) {
...
+    const pendingReleaseChannels = nextState[revision];
+    Object.keys(pendingReleaseChannels).forEach(channel => {
+      const pendingRelease = pendingReleaseChannels[channel];
+
+      if (
+        pendingRelease.progressive &&
+        pendingRelease.progressive.key === key
+      ) {
+        nextState = releaseRevision(state, previousRevision, channel, null);
+        nextState[previousRevision.revision][channel].replaces = pendingRelease;
```
Comment (bartaz) [inline static/js/publisher/release/reducers/pendingReleases.js]: Please add `replaces` to the object schema in the comment below.

File: static/js/publisher/release/reducers/pendingReleases.test.js [hunk 4]
```diff
@@ -776,4 +777,86 @@ describe("pendingReleases", () => {
...
+        const result = pendingReleases(
+          nonProgressiveState,
+          cancelProgressiveReleaseAction
+        );
+
+        expect(result).toEqual(nonProgressiveState);
+      });
+    });
+
+    describe("when there is a progressive pendingRelease", () => {
```
Comment (bartaz) [inline static/js/publisher/release/reducers/pendingReleases.test.js]: I think there is a missing case when there are multiple releases with same progressive key

Comment (Lukewh) [reply static/js/publisher/release/reducers/pendingReleases.test.js]: Added - this also clarifies the expected (if wrong, currently) behaviour a little

File: static/js/publisher/release/reducers/pendingReleases.js (hunk 2 above)
Comment (Lukewh) [reply static/js/publisher/release/reducers/pendingReleases.js]: I've added a comment above this function to explain a little

File: static/js/publisher/release/actions/pendingReleases.test.js [hunk 5]
```diff
@@ -225,4 +227,42 @@ describe("pendingReleases actions", () => {
...
+      channels: ["latest/edge"],
+      confinement: "strict",
+      created_at: "2019-07-16T08:58:04Z",
+      epoch: { read: null, write: null },
+      grade: "stable",
+      revision: 3,
+      "sha3-384": "test",
+      size: 4096,
+      status: "Published",
+      version: "1.8.0"
```
Comment (bartaz) [inline static/js/publisher/release/actions/pendingReleases.test.js]: Wow, providing just an object with id or any necessary fields would be enough, but sure, we can be very specific ;)

**Pull request merged**